Automattic/woocommerce-payments
# Fix double conversion issues on manual subscription renewals.

Fixes #2120

#### Changes proposed in this Pull Request

* Create `Compatibility` class to use for compatibility issues.
* Create `Utils` class to give utilities somewhere to live.
* Created methods to check for subscription renewal in cart, `cart_contains_renewal` and `is_product_subscription_renewal`. These are then used in `should_convert_product_price`.
* `should_convert_product_price` created to determine if the price being passed should be converted, is currently being used in `Frontend_Prices`.
* `override_selected_currency` method create to determine if selected/session currency should be overridden. This needs to happen so that the currency is set to the subscription currency.
* `should_hide_widgets` created to hide widgets if needed, and has been added to current legacy widget.
* `is_call_in_backtrace` method created to check backtrace calls to see if certain functions/methods have been called.
* Update filters that call `get_product_price` to pass 2 arguments, the second being the product.

#### Testing instructions

* Create a subscription product and sign up, if not already done.
* Go to WooCommerce > Subscriptions, and click into a subscription.
* Use _Subscription actions_ at the top right to _Create pending renewal order_ and click the arrow button.
* Navigate to the new order through the link in the _Order notes_.
* The order will have a _Customer payment page_ link above the Status, click that link.
* The checkout page will have the correct price listed.
* Shipping costs and taxes should still convert correctly.
* All other products in the shop should still convert correctly.
* You should be able to add a simple product to the cart with the proper conversion.

-------------------
